cvat-sdk/cvat_sdk/pytorch/__init__.py

```diff
@@ -0,0 +1,280 @@
+import base64
+import collections
+import json
+import os
+import shutil
+import types
+import zipfile
+from concurrent.futures import ThreadPoolExecutor
+from pathlib import Path
+from typing import (
+    Callable,
+    Dict,
+    FrozenSet,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Tuple,
+    Type,
+    TypeVar,
+)
+
+import appdirs
+import attrs
+import attrs.validators
+import PIL.Image
+import torchvision.datasets
+from typing_extensions import TypedDict
+
+import cvat_sdk.core
+import cvat_sdk.core.exceptions
+from cvat_sdk.api_client.model_utils import to_json
+from cvat_sdk.core.utils import atomic_writer
+from cvat_sdk.models import DataMetaRead, LabeledData, LabeledImage, LabeledShape, TaskRead
+
+_ModelType = TypeVar("_ModelType")
+
+_CACHE_DIR = Path(appdirs.user_cache_dir("CVAT_SDK", "CVAT.ai"))
+_NUM_DOWNLOAD_THREADS = 4
+
+
+class UnsupportedDatasetError(cvat_sdk.core.exceptions.CvatSdkException):
+    pass
+
+
+@attrs.frozen
+class FrameAnnotations:
+    tags: List[LabeledImage] = attrs.Factory(list)
+    shapes: List[LabeledShape] = attrs.Factory(list)
+
+
+@attrs.frozen
+class Target:
+    annotations: FrameAnnotations
+    label_id_to_index: Mapping[int, int]
+
+
+class TaskDataset(torchvision.datasets.VisionDataset):
```

Comment:
If we don't inherit the class, which problems we will have?

Reply:
OK. Probably extras_require is the right approach here.

Reply:
But at the same time, if we want to test the feature, we need to install torch and torchvision explicitly, but indeed they don't need us. What do think?

Reply:
> If we don't inherit the class, which problems we will have?

That depends on what the alternative is. The PyTorch [`DataLoader`](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader) class requires the dataset to be an instance of `torch.utils.data.Dataset`, so we would at least have to inherit from that.

We could inherit directly from `Dataset`, rather than `VisionDataset` (in fact, for a long time, that _was_ what my prototype did). The `VisionDataset` class provides dataset transform support that helps the class behave more like other torchvision datasets; however we could, in principle, reimplement that ourselves. I eventually decided to use `VisionDataset` because IMO, if you're working with vision datasets in PyTorch, you'll want to use torchvision anyway for its [image processing transforms](http://pytorch.org/vision/stable/transforms.html).

We could implement multiple classes (e.g. `TaskDataset` inheriting from `Dataset` and `TaskVisionDataset` inheriting from `VisionDataset`) and multiple sets of extras (with torchvision and without), but that just seems like overcomplicating things.

> But at the same time, if we want to test the feature, we need to install torch and torchvision explicitly, but indeed we don't need them. What do think?

I don't understand the question. Naturally we would have to install PyTorch and torchvision to test integration with them.